### PR TITLE
Use chrome-headless instead of PhantomJS for rendering

### PR DESCRIPTION
**Pull request by mauve**
## Summary

It actually works.

## Remaining tasks:

- [x] actually start chromium in headless mode
- [x] bundle chromium (probably easiest to look where puppeteer gets it from)
   - [x] execute bundled chromium
- [x] add some settings (`headless` `on/off` for example, `executable_path`)
- [x] remove PhantomJS bundling and code
- [x] reuse browsers
- [x] add dependency on `chromedp/chromedp`
- [x] show time range and time zone in rendered image
- [ ] add a build-flag for skipping the server side rendering completely
- [ ] only download chromium for supported OS versions

### CODE REVIEW

**Comment by mauve** [comment]
**reminder to self**: the class name for broken for panel errors is `panel-info-corner--error`

**Comment by mauve** [comment]
The `tz` parameter to `/render/` is no longer supported (it was implemented using environment variables before which does no longer work) and I couldn't figure out a simple way of supporting it without changing a lot, therefore I opted to render the timerange and timezone in the picture instead which should be a good compromise:

finally-timerange

**Comment by mauve** [comment]
rebased on upstream/master

File: package.json [hunk 1]
```diff
@@ -129,6 +129,9 @@
     "printWidth": 120
   },
   "license": "Apache-2.0",
+  "grafana": {
+    "chromium_revision": "533271"
```
Comment (torkelo) [inline package.json]: what's the reason for this?

File: build.go [hunk 2]
```diff
@@ -416,6 +420,7 @@ func ldflags() string {
 	b.WriteString(fmt.Sprintf(" -X main.version=%s", version))
 	b.WriteString(fmt.Sprintf(" -X main.commit=%s", getGitSha()))
 	b.WriteString(fmt.Sprintf(" -X main.buildstamp=%d", buildStamp()))
+	b.WriteString(fmt.Sprintf(" -X main.chromiumRevision=%s", chromiumRevision))
```
Comment (torkelo) [inline build.go]: why is this a build flag?

**Review by torkelo — COMMENTED** [review]
How do I get this to work? what do I set chromium revision to?

**Comment by torkelo** [comment]
I get this error "renderer.Init not called"

**Comment by torkelo** [comment]
sorry, had not configured the exec path correctly. Using latest chrome stable I get this error:
""Could not run tasklist: Invalid parameters (-32602)"

Is chromium required?

File: package.json (hunk 1 above)
Comment (mauve) [reply package.json]: We need to have a version we want to bundle, I copied the version which puppeteer is currently using. The reason why it is in `package.json` is because I need to put it somewhere where I can access it from grunt.

File: build.go (hunk 2 above)
Comment (mauve) [reply build.go]: because we need to know the version of the bundled chromium in order to find the correct path

**Comment by mauve** [comment]
chromium needs to be downloaded (it is downloaded automatically by grunt); however when running from master (e.g. `bra run`) the build flag is not included so the exec-path to the chromium executable needs to be configured.

File: public/app/features/panel/panel_header.ts [hunk 3]
```diff
@@ -26,6 +26,12 @@ var template = `
     </ul>
   </span>
   <span class="panel-time-info" ng-show="ctrl.timeInfo"><i class="fa fa-clock-o"></i> {{ctrl.timeInfo}}</span>
+  <span class="panel-time-info" ng-show="ctrl.panel.soloMode">
+    <i class="fa fa-clock-o"></i>
+    {{ctrl.timeRangeString}}
+    <span class="gf-timepicker-utc" ng-show="ctrl.dashboard.isTimezoneUtc()">UTC</span>
+    <span class="gf-timepicker-utc" ng-show="!ctrl.dashboard.isTimezoneUtc()">{{ctrl.dashboard.getTimezone()}}</span>
+  </span>
```
Comment (torkelo) [inline public/app/features/panel/panel_header.ts]: can you remove this change? It seems unrelated to the PR

**Comment by torkelo** [comment]
sorry, I missed the new grunt task that downloaded chromium :)

File: public/app/features/panel/panel_header.ts (hunk 3 above)
Comment (mauve) [reply public/app/features/panel/panel_header.ts]: the previous codebase had a query-parameter `tz` which was mostly set to `UTC` that parameter was sent to phantom using environment variables. Since the chromium process is reused between requests we cannot specify the timezone via environment variable which also means that the `tz` request parameter is removed (until it is supported in `d-solo`).

As the timezone can no longer be specified the rendered image will have the timezone as specified in the dashboard preference and *not* as specified via the request parameter (no longer supported). In order to solve that I opted for just putting the timing information and timezone information into the rendered image.

**Comment by fg2it** [comment]
@mauve @torkelo with the problems that  PhantomJS project seems to have, it is certainly wise to seek an alternative. I would like to highlight several concerns :
- some of us are using grafana on non x86-64 platform. Building it is currently quite easily (PhantomJS dependency is in fact the only real problem) :  with some set up `go run build.go [...] build pkg-deb` works fine to cross build a deb package (at least on armv6/v7/aarch64). This is done with some support inside grafana build process (build.go and grunt task/options).
- chrome has a lot of dependencies, staying "distro agnostic" could be a challenge.
- PhantomJS is certainly needed for server side png rendering but it is probably not the most needed feature. Moving this feature in some kind of plugin available only on x86_64 platform (if too troublesome to have it everywhere) would be better than too much juggling to build on other platform (probably without this feature anyway).

I know arm platform is not the target of grafana but it is certainly used a lot : counting only the packages I produce, 12k downloads on github, 55k on bintray and 36k on docker over the last 2 years. Grafana is great, please let us have the next versions on as much platforms as possible. Thanks!

**Comment by mauve** [comment]
@fg2it do you want to have support for both or do you just want it to be possible to disable the server-side rendering using a build-flag?

**Comment by fg2it** [comment]
@mauve In an ideal world, both (rendering via chrome and a build-flag), but the build flag will probably be much easier to do and failsafe. Moreover, it doesn't prevent from adding rendering support later if this happen to be doable without too much trouble (or even now if I am being too pessimistic). 

An other thing that could help would be to have a build option to supply another chrome version (e.g. version 60 is available on raspbian/stretch for pi).

**Comment by mauve** [comment]
introducing a build flag or even making it an runtime flag should not be too much hassle actually, let me give it a try later this week

File: pkg/components/renderer/renderer.go [hunk 4]
```diff
@@ -1,159 +1,256 @@
...
+	return "", fmt.Errorf("Unexpected OS %s you need to manually set chromium exec path", runtime.GOOS)
+}
+
+func screenshotAction(width, height int, url, renderKey, pngPath string) chromedp.Tasks {
+	var buf []byte
+	return chromedp.Tasks{
+		setViewportAction(width, height),
+		setCookieAction(getLocalDomain(), "renderKey", renderKey),
+		chromedp.Navigate(url),
+		chromedp.WaitReady(`.renderingComplete`, chromedp.ByQuery),
```
Comment (amuraru) [inline pkg/components/renderer/renderer.go]: This will work for single panel rendering only. Please note that currently phantomjs can also do full dashboard rendering /render/full/path/to/dashboard.
Ideally we'd like to keep this support available.

Comment (mauve) [reply pkg/components/renderer/renderer.go]: Ok will fix.

In order for this to work I will add a `.dashboardRenderingComplete` can you point me a place in the code where such an event would be added?

File: scripts/grunt/options/chromium.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    const chromium_revision = grunt.file.readJSON('./package.json').grafana.chromium_revision;
+    const dest = `./tools/chromium/${chromium_revision}`;
+    const downloadURLs = {
+      linux: `https://storage.googleapis.com/chromium-browser-snapshots/Linux_x64/${chromium_revision}/chrome-linux.zip`,
+      mac: `https://storage.googleapis.com/chromium-browser-snapshots/Mac/${chromium_revision}/chrome-mac.zip`,
+      win32: `https://storage.googleapis.com/chromium-browser-snapshots/Win/${chromium_revision}/chrome-win32.zip`,
+      win64: `https://storage.googleapis.com/chromium-browser-snapshots/Win_x64/${chromium_revision}/chrome-win32.zip`,
+    };
+
+    var platform = os.platform();
```
Comment (fg2it) [inline scripts/grunt/options/chromium.js]: As it is, you would end up downloading Linux_x64 version on Linux armv6/v7/aarch64. It would be better to avoid the download.

File: scripts/grunt/options/chromium.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        .pipe(fs.createWriteStream(downloadTarget))
+        .on('error', (err) => {
+          grunt.log.error(`Saving of ${downloadSrc} to ${downloadTarget} failed: ${err}`);
+          done(err);
+        })
+        .on('finish', () => {
+          // unzip
+          grunt.log.writeln(`Unpacking ${downloadTarget} to ${dest}`);
+
+          return extractZip(downloadTarget, dest).then((err) => {
```
Comment (fg2it) [inline scripts/grunt/options/chromium.js]: I didn't fully check how packaging was changed but in `.zip` it seems there is at least one huge file (100Mo) that could be avoided on win (`interactive_ui_tests.exe`) if the name is not as missleading as the one of the `zip`.

File: scripts/grunt/options/chromium.js (hunk 5 above)
Comment (mauve) [reply scripts/grunt/options/chromium.js]: will fix

File: scripts/grunt/options/chromium.js (hunk 6 above)
Comment (mauve) [reply scripts/grunt/options/chromium.js]: unsure how I feel about poking around in the contents of the chromium package.

Comment (fg2it) [reply scripts/grunt/options/chromium.js]: I do agree. I have mix feeling with this. But, it is a hefty binary. Packages will be very large.

**Comment by mauve** [comment]
chromedp is not good enough @torkelo  is trying another approach using puppeeter in PR#10751
